brax/envs/__init__.py

```diff
@@ -30,33 +30,79 @@
 from brax.envs import ur5e
 from brax.envs import wrappers
 
-_envs = {
-    'fetch': fetch.Fetch,
-    'ant': ant.Ant,
-    'grasp': grasp.Grasp,
-    'halfcheetah': halfcheetah.Halfcheetah,
-    'humanoid': humanoid.Humanoid,
-    'ur5e': ur5e.Ur5e,
-    'reacher': reacher.Reacher,
-    'reacherangle': reacherangle.ReacherAngle,
+_envs: Dict[str, Type[env.Env]] = {
+    "fetch": fetch.Fetch,
+    "ant": ant.Ant,
+    "grasp": grasp.Grasp,
+    "halfcheetah": halfcheetah.Halfcheetah,
+    "humanoid": humanoid.Humanoid,
+    "ur5e": ur5e.Ur5e,
+    "reacher": reacher.Reacher,
+    "reacherangle": reacherangle.ReacherAngle,
 }
 State = env.State
 Env = env.Env
 
 
 def create(env_name: str, **kwargs) -> Env:
-  """Creates an Env with a specified brax system."""
-  return _envs[env_name](**kwargs)
+    """Creates an Env with a specified brax system."""
+    return _envs[env_name](**kwargs)
 
 
 def create_fn(env_name: str, **kwargs) -> Callable[..., Env]:
-  """Returns a function that when called, creates an Env."""
-  return functools.partial(create, env_name, **kwargs)
+    """Returns a function that when called, creates an Env."""
+    return functools.partial(create, env_name, **kwargs)
 
 
-def create_gym_env(env_name: str, seed: int = 0, **kwargs) -> gym.Env:
-  environment = create(env_name=env_name, **kwargs)
-  if environment.batch_size:
-    return wrappers.VectorGymWrapper(environment, seed=seed)
-  else:
-    return wrappers.GymWrapper(environment, seed=seed)
+@overload
```

Comment:
Our internal type checking system complained about the `@overload` annotations... I'm not entirely sure why, but I suspect it's something to do with kwargs making the different function signatures ambiguous for matching.

In any case, I think we gain some simplicity by explicitly only returning `gym.Env` as the API, rather than exposing the implementation types `GymWrapper` and `VectorGymWrapper` to the user.

Can you see if your PR works without these overloads, and keeping the old single `create_gym_env(...)-> gym.Env` function signature?  But go ahead and add the `backend` parameter - that is useful, thanks.

Reply:
Would a `Union[gym.Env, gym.vector.VectorEnv]` work?
What kind of error are you getting? Might I ask what type checking system you are using? Is it comparable to mypy? 
It is in my opinion somewhat important for this function to indicate the distinction, since they behave differently, in subtle yet important ways! (e.g. `done` being a vector, the presence of `single_observation_space`, `num_envs`, etc.)

Reply:
Avoiding the `@overload` and returning `Union[gym.Env, gym.vector.VectorEnv]` would fix the specific error I saw, yes. 

Where do you use `single_observation_space` or `num_envs`?  I don't see them accessed anywhere.  This is actually a good thing, as it means you are conforming to the gym API exactly, and that you don't need extra hidden information to process the environment.  If you do need to access them, you can just 

Looking through this code again, I would actually recommend removing them as fields.  And if you do need `num_envs` or `single_observation_space`, you can get both directly from `observation_space`, no?

The concern that `done` changes from a scalar to a vector I'm sympathetic to, but I think the `gym.Env` api already has a rich history of returning very different types depending on how it is instantiated, with no type guarantees.  Unless there's a use case right now that breaks, I think it's OK to follow this same convention.

Reply:
> Avoiding the @overload and returning Union[gym.Env, gym.vector.VectorEnv] would fix the specific error I saw, yes.

What was that error? Would you be able to share the error message you were getting?

> Where do you use single_observation_space or num_envs? I don't see them accessed anywhere. This is actually a good thing, as it means you are conforming to the gym API exactly, and that you don't need extra hidden information to process the environment.

What i mean is that it's important for the user code to be able to tell (at static type-checking time) whether the returned value is a `gym.Env` instance, or a `gym.vector.VectorEnv` instance, because if it's a `gym.vector.VectorEnv`, then it behaves differently than a `gym.Env`.
Sorry I wasn't clear before.

Reply:
There error was from using pytype:

```
File "brax/envs/__init__.py", line 78, in create_gym_env: bad return type [bad-return-type]
           Expected: gym.vector.vector_env.VectorEnv
  Actually returned: brax.envs.wrappers.GymWrapper
```

See: https://google.github.io/pytype/errors.html#bad-return-type

Why is it important for the user code to distinguish between these two types statically?  I hear you that the implementations are different, but what matters is that the API of the two classes is exactly the same.  This is exactly why `gym.vector.VectorEnv` inherits from `gym.Env`, so that users can just accept `gym.Env` and move on.  What specific use case are you seeing that is breaking otherwise?

Reply:
Thanks for giving the error, I'll take a look and try to find a fix.

Yes, VectorEnv inherits from gym.Env, but that doesnt mean that the APIs are exactly the same, it just means that there is overlap!

Yes there are plenty of differences which are very important to be aware of from the user side! Here is but a few:

- Done is a vector
- (currently) no rendering support
- final observation is wasted in VectorEnvs
- VectorEnvs can't (currently) be vectorized

Since you can't (and shouldn't) treat aVectorEnv *exactly* like a gym.Env, it would be very good, in my opinion, that if you use a single function for both, that the static type checking can at least inform users of which one they are going to get.

If you really can't tolerate the typing error, then Id suggest reverting back to two different functions, one for a gym.Env, and another for a VectorEnv.

Returning Union, while it appears to be a bit of a middle-ground, actually makes user code uglier and more complicated, because it shifts the responsibility of checking the return type to the user side, using either asserts or isinstance checks in user code, which isnt great.

Reply:
By the way I think the error might be caused by the value batch_size=0, which is an int, so the type check expects a VectorEnv, but the if clause returns a Env instead!

I'll try a few things tomorrow, but Im pretty confident that I'll be able to get that error to disapear.